src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+
+import "../interfaces/IBLSPubkeyRegistry.sol";
+import "../interfaces/IRegistryCoordinator.sol";
+import "../interfaces/IBLSPublicKeyCompendium.sol";
+
+import "../libraries/BN254.sol";
+
+import "forge-std/Test.sol";
+
+
+contract BLSPubkeyRegistry is IBLSPubkeyRegistry, Test {
+    using BN254 for BN254.G1Point;
+
+    /// @notice the hash of the zero pubkey aka BN254.G1Point(0,0)
+    bytes32 internal constant ZERO_PK_HASH = hex"ad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5";
+    /// @notice the current global aggregate pubkey among all of the quorums
+    BN254.G1Point public globalApk;
+    /// @notice the registry coordinator contract
+    IRegistryCoordinator public registryCoordinator;
+    /// @notice the BLSPublicKeyCompendium contract against which pubkey ownership is checked
+    IBLSPublicKeyCompendium public immutable pubkeyCompendium;
+
+    // list of all updates made to the global aggregate pubkey
+    ApkUpdate[] public globalApkUpdates;
+    // mapping of quorumNumber => ApkUpdate[], tracking the aggregate pubkey updates of every quorum
+    mapping(uint8 => ApkUpdate[]) public quorumApkUpdates;
+    // mapping of quorumNumber => current aggregate pubkey of quorum
+    mapping(uint8 => BN254.G1Point) private quorumApk;
+
+    event PubkeyAdded(
+        address operator,
+        BN254.G1Point pubkey
+    );
+    event PubkeyRemoved(
+        address operator,
+        BN254.G1Point pubkey
+    );  
+
+    modifier onlyRegistryCoordinator() {
+        require(msg.sender == address(registryCoordinator), "BLSPubkeyRegistry.onlyRegistryCoordinator: caller is not the registry coordinator");
+        _;
+    }
+
+    constructor(
+        IRegistryCoordinator _registryCoordinator,
+        IBLSPublicKeyCompendium _pubkeyCompendium
+    ){
+        registryCoordinator = _registryCoordinator;
+        pubkeyCompendium = _pubkeyCompendium;
+    }
```

Comment:
should perhaps have a wider discussion on if this + other contracts should be upgradeable (or support both upgradeable and non-upgradeable uses?)

Reply:
definitely

Reply:
agreed. I think the pubkeyCompendium should be non-upgradeable because we anticipate many middlewares will use it and it is siloed logic. for the other contracts, we should have more discussion

Reply:
I agree about the PubkeyCompendium
I think the other contracts should be designed + built to support upgradeable and non-upgradeable versions, but I'm OK merging this and making that change later.